src/proxy/http/add_header.rs

```diff
@@ -0,0 +1,229 @@
+use std::{fmt, marker::PhantomData};
+
+use http::header::{AsHeaderName, HeaderValue};
+
+use svc;
+
+type Ret<T> = fn(&T) -> Option<HeaderValue>;
+
+/// Wraps HTTP `Service` `Stack<T>`s so that a given header is removed from a
+/// request or response.
+#[derive(Clone)]
+pub struct Layer<H, T, R> {
+    header: H,
+    retrieve: Ret<T>,
+    _req_or_res: PhantomData<fn(R)>,
+}
+
+/// Wraps an HTTP `Service` so that a given header is added from each request
+/// or response.
+#[derive(Clone)]
+pub struct Stack<H, T, M, R> {
+    header: H,
+    retrieve: Ret<T>,
+    inner: M,
+    _req_or_res: PhantomData<fn(R)>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Service<H, S, R> {
+    header: H,
+    value: HeaderValue,
+    inner: S,
+    _req_or_res: PhantomData<fn(R)>,
+}
+
+// === impl Layer ===
+
+/// Call `request::layer(header)` or `response::layer(header)`.
+fn layer<H, T, R>(header: H, retrieve: Ret<T>) -> Layer<H, T, R>
+where
+    H: AsHeaderName + Clone,
+    R: Clone,
+{
+    Layer {
+        header,
+        retrieve,
+        _req_or_res: PhantomData,
+    }
+}
+
+impl<H, T, M, R> svc::Layer<T, T, M> for Layer<H, T, R>
+where
+    H: AsHeaderName + Clone + fmt::Debug,
+    T: fmt::Debug,
+    M: svc::Stack<T>,
+{
+    type Value = <Stack<H, T, M, R> as svc::Stack<T>>::Value;
+    type Error = <Stack<H, T, M, R> as svc::Stack<T>>::Error;
+    type Stack = Stack<H, T, M, R>;
+
+    fn bind(&self, inner: M) -> Self::Stack {
+        Stack {
+            header: self.header.clone(),
+            retrieve: self.retrieve,
+            inner,
+            _req_or_res: PhantomData,
+        }
+    }
+}
+
+impl<H, T, R> fmt::Debug for Layer<H, T, R>
```

Comment:
Can we get this to display whether the layer is a request layer or a response layer?

I *think* you can do something like this:

```rust
impl<H, T> fmt::Debug for Layer<H, T, ReqHeader>
where
    H: fmt::Debug,
    T: fmt::Debug,
{
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_struct("Layer")
            .field("header", &self.header)
            .field("retrieve", &format_args!("{}", "..."))
            .field("req_or_res", &format_args!("{}", "Request"))
            .finish()
    }
}


impl<H, T> fmt::Debug for Layer<H, T, ResHeader> 
where
    ...
```

Reply:
When do we expect these layers to be logged?  they really only exist in main.rs ephemerally...

i'd rather keep our boilerplate code to a minimum if there's no upsid

Reply:
@olix0r that's fair, I suppose it's really the _service_ that ought to be able to indicate this. Maybe consider replacing its derived `Debug` impl, and remove this one entirely since we're not currently logging layers?

Reply:
@hawkw Are you suggesting `ReqHeader` and `ResHeader` `Debug` impls, but for `Service` instead?

Reply:
@kleimkuhler yeah, that's correct --- I think the `Service` types _may_ on occasion show up in the logs though I'm not sure if that's the case or not.